### ISSUE TITLE
LinearGrids with same nodes have different spacing/weights

### ISSUE THREAD

**Issue by f0uriest** [PlasmaControl/DESC#277]
There are 2 methods for creating a grid with a specified number of nodes (as opposed to specified basis resolution). The first (which isn't officially documented) is to just pass integers for `theta`, `zeta` etc, which will linearly space that many nodes, as in `grid1` below. The other is to pass in arrays directly, as is done with `grid2` below.
```
ntheta = 5
nzeta = 4
NFP = 3
grid1 = LinearGrid(rho=1, theta=ntheta, zeta=nzeta, NFP=NFP, endpoint=False)
grid2 = LinearGrid(rho=1, theta=np.linspace(0,2*np.pi,ntheta, endpoint=False), zeta=np.linspace(0,2*np.pi/NFP, nzeta, endpoint=False), NFP=NFP)

print("Nodes agree:", np.allclose(grid1.nodes, grid2.nodes))
print("Spacing agree:", np.allclose(grid1.spacing, grid2.spacing))
```
Gives that the nodes agree but spacing doesn't. 

The spacing for `grid1` seems correct, as it has all nodes having the same spacing and weights. While the spacing for `grid2` seems way off, with `dz` ranging from 0.2 to 4.9

This further causes the weights to be computed incorrectly, possibly giving the wrong averages etc.

**Comment by dpanici** [comment]
![image](https://user-images.githubusercontent.com/37969854/192907536-c9799d59-88d2-4b4f-83fe-911e572a60b0.png)

this is the culprit, the spacing for when it is not given an int is calculated as if it is a finite difference. To be honest I do not understand what this spacing is, if we just want spacing would we not just do 
```dt[0:-1] = t[1:] - t[0:-1]
dt[-1] = 2pi - t[-1]
```
We want spacing to add up to 1 for dr and 2pi for dt and dz. 
- If it is passed in a linearly spaced set of points in any of the coordiantes, the spacing should be uniform
 - this could be easily accomplished by subtracting the second point in the array from the first
 - however this fails for an array that is NOT linearly spaced
  - this begs the question, should we allow such a spacing in the `LinearGrid` class? it seems the logic here is trying to come up with some sort of sensible spacing/weighting for non-uniformly spaced grid points, but maybe those should be put in the `Grid` class and not the `LinearGrid` class anyways

**Comment by f0uriest** [comment]
I think something like `np.diff` could work for a more general case. Just need to be careful with theta and zeta because they're periodic.

**Comment by unalmis** [comment]
I've pushed a commit that resolves the original issue Rory raised opened. There is still a discrepancy when `endpoint = True`. More on that at the bottom.


I'm not certain why, but we changed the linear grid spacing computations from `sup = 2pi or 1`, `dx = sup / (max(1, x.size))` to the for loop version of this:
```
if x.size > 1:
    dx[0] = (x[0] + x[1]) / 2
    dx[1:-1] = (x[2:] - x[:-2]) / 2
    dx[-1] = sup - (x[-2] + x[-1]) / 2
else:
    dx = 2pi
```

Couple separate issues with the above algorithm:
-  `dx[0]` and `dx[-1]` are not computed correctly (at least not when `endpoint = False`).  Changing the following fixes that.
    - `dx[0] = x[1] - x[0]`
    - `dx[-1] = sup - x[-1]` 
- it does not compute dzeta correctly. While the zeta nodes depend on NFP, the zeta spacing should not. Fixed by just scaling dz by NFP.

The middle `dx[1:-1]` computation is fine. I'm not sure why it was written that way, as opposed to `dx[0:-1] = np.diff(x)`. Maybe an attempt to smooth out non-uniform spacing.


For the endpoint issue:
Adding an endpoint at 2pi will make `dx[-1] = 2pi - x[-1] = 0` when you pass an array. When you pass an integer, it's not zero (see Dario's screen shot).
Since the point of `endpoint=True` is to duplicate the `theta=0` node at `theta=2pi`, do we want their spacing to be the same?

**Comment by f0uriest** [comment]
the zeta spacing should depend on NFP, since the nodes only go between 0 and 2pi/NFP. I think all you'd need to do is set `sup=2pi/NFP` for zeta. Including an endpoint still does seem to be an issue, I'm not sure of the best fix. In some way it doesn't really make sense since you then have 2 nodes at the same point, assuming you're always doing `linspace(0,sup,n)`. Having `endpoint=True` only really makes sense if you're not going all the way out to `sup`. 
One counterpoint though is that in some of the plotting routines we probably need `endpoint=True` to get stuff to properly close on themselves or plot all the way to the edge of the axis. However, for plotting we don't care about `spacing` or `weights` so it's probably ok if they're "wrong" in that case.

**Comment by unalmis** [comment]
> the zeta spacing should depend on NFP
I think @ddudt and I decided that the spacing shouldn't correspond to the coordinate of a node. Instead it should correspond to the weight of a node. For example, we want `Y = sum(dzeta * something)` to not be dependent on NFP. If they do depend on NFP, then a surface average would give different results for NFP = 1 and NFP = 2.

**Comment by ddudt** [comment]
> > the zeta spacing should depend on NFP
> 
> I think @ddudt and I decided that the spacing shouldn't correspond to the coordinate of a node. Instead it should correspond to the weight of a node. For example, we want `Y = sum(dzeta * some_array)` to not be dependent on NFP. If `dzeta` depends on NFP, then a surface average, which uses integrals like these, would give different results for NFP = 1 and NFP = 2.
> 
> (~An alternative is to always do `Y = sum(dzeta / NFP * some_array)`~ actually I remember that only works for grids on a single surface grid. I had latex'd something describing that a couple months ago if you want to see)
> 
> (By the way, in your example, grid1's spacing is [independent of NFP.](https://github.com/PlasmaControl/DESC/blob/ec7f38b8f2423a382700ac1c6a6b6e70660873e7/desc/grid.py#L467) While grid2's was dependent on NFP)

Correct. Since we compute the weights based on the spacing, I think we want to scale the spacing by NFP and not use the true spacing between the zeta coordinates.
